src/coreComponents/physicsSolvers/wavePropagation/AcousticWaveEquationSEMKernel.hpp

```diff
@@ -0,0 +1,500 @@
+    {
+      constexpr localIndex numNodesPerElem = FE_TYPE::numNodes;
+      real64 xLocal[numNodesPerElem][3];
+      for( localIndex a = 0; a < numNodesPerElem; ++a )
+      {
+        for( localIndex i = 0; i < 3; ++i )
+        {
+          xLocal[a][i] = X( elemsToNodes( elementIndex, a ), i );
+        }
+      }
+
+      // coordsOnRefElem = invJ*(coords-coordsNode_0)
+      localIndex const q = 0;
+
+      real64 invJ[3][3] = {{ 0 }};
+      FE_TYPE::invJacobianTransformation( q, xLocal, invJ );
+
+      real64 coordsRef[3] = { 0 };
+      for( localIndex i = 0; i < 3; ++i )
+      {
+        coordsRef[i] = coords[i] - xLocal[q][i];
+      }
+
+      for( localIndex i = 0; i < 3; ++i )
+      {
+        // Init at (-1,-1,-1) as the origin of the referential elem
+        coordsOnRefElem[i] = -1.0;
+        for( localIndex j = 0; j < 3; ++j )
+        {
+          coordsOnRefElem[i] += invJ[i][j]*coordsRef[j];
+        }
+      }
+      return true;
+    }
+    else
+    {
+      return false;
+    }
+  }
+
+  /**
+   * @brief Launches the precomputation of the source and receiver terms
+   * @param size the number of cells in the subRegion
+   * @param numNodesPerElem number of nodes per element
+   * @param X coordinates of the nodes
+   * @param elemsToNodes map from element to nodes
+   * @param faceNodeIndices cell-wise storage for the nodes
+   * @param sourceCoordinates coordinates of the source terms
+   * @param sourceIsLocal flag indicating whether the source is local or not
+   * @param sourceNodeIds indices of the nodes of the element where the source is located
+   * @param sourceNodeConstants constant part of the source terms
+   * @param receiverCoordinates coordinates of the receiver terms
+   * @param receiverIsLocal flag indicating whether the receiver is local or not
+   * @param receiverNodeIds indices of the nodes of the element where the receiver is located
+   * @param receiverNodeConstants constant part of the receiver term
+   */
+  template< typename EXEC_POLICY, typename FE_TYPE >
+  static void
+  launch( localIndex const size,
+          localIndex const numNodesPerElem,
+          arrayView2d< real64 const, nodes::REFERENCE_POSITION_USD > const X,
+          arrayView2d< localIndex const, cells::NODE_MAP_USD > const & elemsToNodes,
+          arrayView1d< arrayView1d< arrayView1d< localIndex > const > const > const allFaceNodeIndices,
+          arrayView2d< real64 const > const sourceCoordinates,
+          arrayView1d< localIndex > const sourceIsLocal,
+          arrayView2d< localIndex > const sourceNodeIds,
+          arrayView2d< real64 > const sourceConstants,
+          arrayView2d< real64 const > const receiverCoordinates,
+          arrayView1d< localIndex > const receiverIsLocal,
+          arrayView2d< localIndex > const receiverNodeIds,
+          arrayView2d< real64 > const receiverConstants )
+  {
+
+    forAll< EXEC_POLICY >( size, [=] GEOSX_HOST_DEVICE ( localIndex const k )
```

Comment:
Just so I understand, this is looping over every element, and then for each element it checks if each source/receiver is contained inside?